File: src/lib/deck.ts
import type { ChallengeCard, Rarity } from "@/types/card";
import { RARITY_CONFIG } from "@/data/rarityConfig";
import { DRAW_COUNT } from "@/types/game";

// === 가중 랜덤 선택 (Weighted Random Selection) ===
// 각 등급에 "무게"를 부여해서, 무게가 큰 등급(normal)이 더 자주 뽑히게 하는 방식
// 주사위를 던진다고 생각하면: normal은 면이 60개, legend는 면이 3개인 주사위

function getWeightedRandomRarity(): Rarity {
  const rarities = Object.entries(RARITY_CONFIG) as [Rarity, typeof RARITY_CONFIG.normal][];
  const totalWeight = rarities.reduce((sum, [, config]) => sum + config.weight, 0);
  let random = Math.random() * totalWeight;

  for (const [rarity, config] of rarities) {
    random -= config.weight;
    if (random <= 0) return rarity;
  }

  return "normal"; // fallback
}

// === 카드 드로우 알고리즘 ===
// 1. 해금된 카드 풀에서 등급별로 가중 랜덤 선택
// 2. 같은 카테고리 중복 최소화 (다양성)
// 3. 피티 시스템: 6장 중 최소 1장은 normal이 아닌 카드 보장

export function drawCards(unlockedCards: ChallengeCard[]): ChallengeCard[] {
  if (unlockedCards.length <= DRAW_COUNT) {
    return shuffleArray([...unlockedCards]);
  }

  const drawn: ChallengeCard[] = [];
  const usedIds = new Set<string>();
  const usedCategories = new Set<string>();
  let attempts = 0;
  const maxAttempts = 100;

  while (drawn.length < DRAW_COUNT && attempts < maxAttempts) {
    attempts++;
    const targetRarity = getWeightedRandomRarity();

    // 해당 등급 카드들 중 아직 안 뽑힌 것들 필터링
    let candidates = unlockedCards.filter(
      (card) => card.rarity === targetRarity && !usedIds.has(card.id)
    );

    // 카테고리 다양성: 이미 뽑힌 카테고리는 피하기 (가능한 경우)
    if (drawn.length < 4) {
      const diverseCandidates = candidates.filter(
        (card) => !usedCategories.has(card.category)
      );
      if (diverseCandidates.length > 0) {
        candidates = diverseCandidates;
      }
    }

    // 해당 등급에 후보가 없으면 다음 시도
    if (candidates.length === 0) continue;

    // 랜덤으로 하나 선택
    const randomIndex = Math.floor(Math.random() * candidates.length);
    const selected = candidates[randomIndex];
    drawn.push(selected);
    usedIds.add(selected.id);
    usedCategories.add(selected.category);
  }

  // 피티 시스템 (Pity System): 6장이 모두 normal이면 하나를 교체
  const hasNonNormal = drawn.some((card) => card.rarity !== "normal");
  if (!hasNonNormal) {
    const nonNormalCards = unlockedCards.filter(
      (card) => card.rarity !== "normal" && !usedIds.has(card.id)
    );
    if (nonNormalCards.length > 0) {
      const replacement = nonNormalCards[Math.floor(Math.random() * nonNormalCards.length)];
      drawn[drawn.length - 1] = replacement;
    }
  }

  return drawn;
}

// === 등급 가중치 기반 N장 드로우 (카드팩 등 범용) ===
// 중복 없이, 등급 가중치에 따라 pool에서 count장 선택
export function drawFromPool(pool: ChallengeCard[], count: number): ChallengeCard[] {
  if (pool.length <= count) return shuffleArray([...pool]);

  const result: ChallengeCard[] = [];
  const usedIds = new Set<string>();
  let attempts = 0;

  while (result.length < count && attempts < 100) {
    attempts++;
    const targetRarity = getWeightedRandomRarity();
    const candidates = pool.filter(
      (c) => c.rarity === targetRarity && !usedIds.has(c.id)
    );
    if (candidates.length === 0) continue;

    const selected = candidates[Math.floor(Math.random() * candidates.length)];
    result.push(selected);
    usedIds.add(selected.id);
  }

  // 피티: 전부 normal이면 1장 교체
  if (result.length > 0 && result.every((c) => c.rarity === "normal")) {
    const nonNormal = pool.filter((c) => c.rarity !== "normal" && !usedIds.has(c.id));
    if (nonNormal.length > 0) {
      result[result.length - 1] = nonNormal[Math.floor(Math.random() * nonNormal.length)];
    }
  }

  return result;
}

// 배열 섞기 (Fisher-Yates Shuffle)
function shuffleArray<T>(array: T[]): T[] {
  for (let i = array.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [array[i], array[j]] = [array[j], array[i]];
  }
  return array;
}

File: upnext-ios/UpNext/UpNext/Models/Deck.swift
//
//  Deck.swift
//  UpNext 로직 — 카드 드로우 알고리즘.
//
//  웹 src/lib/deck.ts + src/data/rarityConfig.ts(weight 부분)를 1:1 포팅.
//
//  비결정론적 (Math.random 사용) — 테스트 벡터 동치성 불가능.
//  대신 알고리즘 구조(가중치/다양성/피티)가 동일함을 보장.
//
//  Phase 2.3 (결정론적 알고리즘 Swift 포팅) 산출물.
//

import Foundation

enum Deck {

    /// 등급별 드로우 가중치. 웹 RARITY_CONFIG.weight.
    /// normal이 가장 자주, legend가 가장 드물게 (60 : 25 : 12 : 3).
    static func drawWeight(_ rarity: Rarity) -> Int {
        switch rarity {
        case .normal: return 60
        case .rare:   return 25
        case .unique: return 12
        case .legend: return 3
        }
    }

    /// 가중 랜덤 등급 선택. 웹 getWeightedRandomRarity.
    /// Rarity.allCases는 normal→rare→unique→legend 순 — 웹 RARITY_CONFIG 객체 순서와 동일.
    static func weightedRandomRarity() -> Rarity {
        let total = Rarity.allCases.reduce(0) { $0 + drawWeight($1) }
        var random = Double.random(in: 0..<1) * Double(total)
        for rarity in Rarity.allCases {
            random -= Double(drawWeight(rarity))
            if random <= 0 { return rarity }
        }
        return .normal  // fallback
    }

    /// 일일 카드 드로우 — 6장. 웹 drawCards.
    ///
    /// 알고리즘:
    ///  1. 해금 풀에서 등급 가중 랜덤 선택
    ///  2. 4장째까지는 카테고리 중복 회피 (다양성)
    ///  3. 피티: 6장이 전부 normal이면 마지막 1장을 non-normal로 교체
    static func drawCards(unlocked: [ChallengeCard]) -> [ChallengeCard] {
        if unlocked.count <= GameConstants.drawCount {
            return unlocked.shuffled()
        }

        var drawn: [ChallengeCard] = []
        var usedIds = Set<String>()
        var usedCategories = Set<Category>()
        var attempts = 0

        while drawn.count < GameConstants.drawCount && attempts < 100 {
            attempts += 1
            let targetRarity = weightedRandomRarity()

            var candidates = unlocked.filter {
                $0.rarity == targetRarity && !usedIds.contains($0.id)
            }
            // 카테고리 다양성 — 4장째까지만 적용
            if drawn.count < 4 {
                let diverse = candidates.filter { !usedCategories.contains($0.category) }
                if !diverse.isEmpty { candidates = diverse }
            }
            if candidates.isEmpty { continue }

            let selected = candidates[Int.random(in: 0..<candidates.count)]
            drawn.append(selected)
            usedIds.insert(selected.id)
            usedCategories.insert(selected.category)
        }

        // 피티 시스템 — 전부 normal이면 마지막 1장 교체
        // (웹은 빈 배열에 drawn[-1] 할당 시 무해하지만 Swift는 크래시 → isEmpty 가드)
        if !drawn.isEmpty && !drawn.contains(where: { $0.rarity != .normal }) {
            let nonNormal = unlocked.filter {
                $0.rarity != .normal && !usedIds.contains($0.id)
            }
            if let replacement = nonNormal.randomElement() {
                drawn[drawn.count - 1] = replacement
            }
        }
        return drawn
    }

    /// 범용 풀 드로우 — count장 (카드팩 등). 웹 drawFromPool.
    static func drawFromPool(_ pool: [ChallengeCard], count: Int) -> [ChallengeCard] {
        if pool.count <= count { return pool.shuffled() }

        var result: [ChallengeCard] = []
        var usedIds = Set<String>()
        var attempts = 0

        while result.count < count && attempts < 100 {
            attempts += 1
            let targetRarity = weightedRandomRarity()
            let candidates = pool.filter {
                $0.rarity == targetRarity && !usedIds.contains($0.id)
            }
            if candidates.isEmpty { continue }

            let selected = candidates[Int.random(in: 0..<candidates.count)]
            result.append(selected)
            usedIds.insert(selected.id)
        }

        // 피티 — 전부 normal이면 마지막 1장 교체
        if !result.isEmpty && result.allSatisfy({ $0.rarity == .normal }) {
            let nonNormal = pool.filter {
                $0.rarity != .normal && !usedIds.contains($0.id)
            }
            if let replacement = nonNormal.randomElement() {
                result[result.count - 1] = replacement
            }
        }
        return result
    }
}
